samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaConsumerProxy.java

```diff
@@ -0,0 +1,447 @@
+      // If the message we are about to consume is < end offset, we are starting with a lag.
+      long initialLag = endOffsets.get(tp) - startingOffset;
+
+      LOG.info("Initial lag for SSP {} is {} (end={}, startOffset={})", ssp, initialLag, endOffsets.get(tp), startingOffset);
+      latestLags.put(ssp, initialLag);
+      sink.setIsAtHighWatermark(ssp, initialLag == 0);
+    });
+
+    // initialize lag metrics
+    refreshLatencyMetrics();
+  }
+
+  // the actual polling of the messages from kafka
+  public Map<SystemStreamPartition, List<IncomingMessageEnvelope>> pollConsumer(
+      Set<SystemStreamPartition> systemStreamPartitions, long timeout) {
+
+    if (topicPartitions2SSP.size() == 0) {
+      throw new SamzaException("cannot poll empty set of TopicPartitions");
+    }
+
+    // Since we need to poll only from some subset of TopicPartitions (passed as the argument),
+    // we need to pause the rest.
+    List<TopicPartition> topicPartitionsToPause = new ArrayList<>();
+    List<TopicPartition> topicPartitionsToPoll = new ArrayList<>();
+
+    for (Map.Entry<TopicPartition, SystemStreamPartition> e : topicPartitions2SSP.entrySet()) {
+      TopicPartition tp = e.getKey();
+      SystemStreamPartition ssp = e.getValue();
+      if (systemStreamPartitions.contains(ssp)) {
+        topicPartitionsToPoll.add(tp);  // consume
+      } else {
+        topicPartitionsToPause.add(tp); // ignore
+      }
+    }
+
+    ConsumerRecords<K, V> records;
+    // make a call on the client
+    try {
+      // Currently, when doing checkpoint we are making a safeOffset request through this client, thus we need to synchronize
+      // them. In the future we may use this client for the actually checkpointing.
+      synchronized (kafkaConsumer) {
+        // Since we are not polling from ALL the subscribed topics, so we need to "change" the subscription temporarily
+        kafkaConsumer.pause(topicPartitionsToPause);
+        kafkaConsumer.resume(topicPartitionsToPoll);
+        records = kafkaConsumer.poll(timeout);
+        // resume original set of subscription - may be required for checkpointing
+        kafkaConsumer.resume(topicPartitionsToPause);
+      }
+    } catch (InvalidOffsetException e) {
+      // If the consumer has thrown this exception it means that auto reset is not set for this consumer.
+      // So we just rethrow.
+      LOG.error("Caught InvalidOffsetException in pollConsumer", e);
+      throw e;
+    } catch (KafkaException e) {
+      // we may get InvalidOffsetException | AuthorizationException | KafkaException exceptions,
+      // but we still just rethrow, and log it up the stack.
+      LOG.error("Caught a Kafka exception in pollConsumer", e);
+      throw e;
+    }
+
+    return processResults(records);
+  }
+
+  private Map<SystemStreamPartition, List<IncomingMessageEnvelope>> processResults(ConsumerRecords<K, V> records) {
+    if (records == null) {
```

Comment:
Just curious, in what scenarios does the records parameter will be null. It appears to me that this check is unnecessary and can be deleted.

Reply:
> if Kafka decides throw an exception we don't handle.


In that scenario, it will be propogated back to the caller, control-flow won't even reach the processResults method here.

Reply:
I guess you are right, but this is just a good coding style - check function's arguments :)